File: web/app/services/media/dtos/media-dtos.ts
import type { PageQuery } from "@/types/global/paginated";

export type MediaKind = "Image" | "Audio" | "Video";
export type MediaStatus = "Pending" | "Confirmed";

/** Request body for POST /teacher/media/presign. */
export interface PresignRequest {
    fileName: string;
    contentType: string;
    byteSize: number;
    width?: number | null;
    height?: number | null;
    durationSeconds?: number | null;
}

/** Result of a presign — where to PUT the bytes + the id to confirm. */
export interface PresignResultDto {
    mediaPublicId: string;
    uploadUrl: string;
    httpMethod: string;
    headers: Record<string, string>;
    expiresAt: string;
}

/** A confirmed/pending media asset (never exposes the internal storage key). */
export interface MediaAssetDto {
    publicId: string;
    url: string;
    kind: MediaKind;
    contentType: string;
    byteSize: number;
    width?: number | null;
    height?: number | null;
    durationSeconds?: number | null;
    status: MediaStatus;
    createdAt: string;
}

/** A media row for the library / admin lists (adds owner for the admin surface). */
export interface MediaListItem extends MediaAssetDto {
    ownerPublicId: string;
    ownerName: string;
}

export interface StorageUsageDto {
    usedBytes: number;
    limitBytes: number; // 0 = unlimited
    isPro: boolean;
    planName: string;
}

export interface StorageOverviewItem {
    ownerPublicId: string;
    ownerName: string;
    totalBytes: number;
    assetCount: number;
}

export interface MediaListQuery extends PageQuery {
    kind?: MediaKind;
    status?: MediaStatus;
}

export interface AdminMediaListQuery extends MediaListQuery {
    ownerPublicId?: string;
}

File: api/src/ClassManagement.Application/Modules/Media/DTOs/MediaDtos.cs
namespace ClassManagement.Application.Modules.Media.DTOs;

// Response DTOs for the Media module (MVP-9). The internal storage_key is NEVER exposed (BR-9-01) — only
// public_id + the CDN url.

/// <summary>Result of a presign request: where to PUT the bytes + the id to confirm afterwards.</summary>
public sealed record PresignResultDto(
    Guid MediaPublicId,
    string UploadUrl,
    string HttpMethod,
    IReadOnlyDictionary<string, string> Headers,
    DateTime ExpiresAt
);

/// <summary>A single media asset (detail / confirm response).</summary>
public sealed record MediaAssetDto(
    Guid PublicId,
    string Url,
    string Kind,
    string ContentType,
    long ByteSize,
    int? Width,
    int? Height,
    int? DurationSeconds,
    string Status,
    DateTime CreatedAt
);

/// <summary>A media asset row for the library / admin lists (adds owner for the admin surface).</summary>
public sealed record MediaListDto(
    Guid PublicId,
    string Url,
    string Kind,
    string ContentType,
    long ByteSize,
    int? Width,
    int? Height,
    int? DurationSeconds,
    string Status,
    DateTime CreatedAt,
    Guid OwnerPublicId,
    string OwnerName
);

/// <summary>Teacher's current storage usage vs. their plan cap (0 = unlimited).</summary>
public sealed record StorageUsageDto(long UsedBytes, long LimitBytes, bool IsPro, string PlanName);

/// <summary>Per-teacher storage aggregate for the admin overview.</summary>
public sealed record StorageOverviewDto(
    Guid OwnerPublicId,
    string OwnerName,
    long TotalBytes,
    int AssetCount
);
